TheJumpCloud/support
# SA-2543: Parallel Implementation

## Issues
* [SA-2543](https://jumpcloud.atlassian.net/browse/SA-2543) - Use Get-JCResults in applicable Get-JC* public functions to make API calls

## What does this solve?
This will allow users to make use of PowerShell 7+ Parallel functionality natively within the module for the following functions:
* **Get-JCUser**
* **Get-JCSystem**
* **Get-JCSystemUser**
* **Get-JCSystemGroupMember**
* **Get-JCUserGroupMember** 

This is particularly beneficial for customers that are querying large quantities of data. Making use of parallel processing will potentially cause functions to complete **_up to_** 60% faster. As an experiment, I measured the time to complete for each of the functions mentioned above in parallel vs in serial. 

The organization that the tests were performed on has excess of 4,000 users and over 20,000 devices.
<p align="center">
<img src="https://user-images.githubusercontent.com/89030113/172896805-358be76c-a7ea-4439-9e8c-5c27ea8d27db.png">
<img src="https://user-images.githubusercontent.com/89030113/172896803-ba2ae0d6-e9b7-46b7-8eba-9e5428a5388c.png">

> :bulb: **Tip:** For more fun graphs, hit the drop down below!

<details>
<summary> Individual function comparisons </summary>



### Get-JCUser
<p align="center">
<img src="https://user-images.githubusercontent.com/89030113/172896801-b8b7a983-8026-4bc4-941b-53479224b3f1.png">
</p>

### Get-JCSystem
<p align="center">
<img src="https://user-images.githubusercontent.com/89030113/172896800-b9d9ebd0-4e6c-4ed5-b177-aca9ab71080a.png">
</p>

### Get-JCSystemUser
<p align="center">
<img src="https://user-images.githubusercontent.com/89030113/172896797-d2c9a067-e709-4cd6-98cb-76c0a66ca576.png">
</p>

### Get-JCSystemGroupMember
<p align="center">
<img src="https://user-images.githubusercontent.com/89030113/172896795-fbcb71df-78bd-4cdf-aa45-cc477d315813.png">
</p>

### Get-JCUserGroupMember
> :memo: **Note:** There was a transient error during pass 24 that caused a longer than expected time to completion.
<p align="center">
<img src="https://user-images.githubusercontent.com/89030113/172896794-40b1c1a8-6be1-4d09-9670-b6b416521e9f.png">
</p>
</details>
</p>

## Is there anything particularly tricky?
There are a few limitations with running these functions in parallel. The two basic ones are the following:
* PowerShell version 7+ installed
* Device running module has multiple cores

> :warning: **Warning:** If those two requirements are not met, the functions will not run in parallel or will not reduce the time to completion as much as anticipated.

Another consideration that needs to be taken into account is the size of the organization and which endpoints are being called. Here are the functions requirements for amount of users/devices in order to successfully run in parallel:
* Get-JCUser: 1001+ Users
    * [Using /search/ endpoint](https://docs.jumpcloud.com/api/1.0/index.html#operation/search_systemusers_post)
* Get-JCSystem: 1001+ Systems
    * [Using /search/ endpoint](https://docs.jumpcloud.com/api/1.0/index.html#operation/search_system_post)
* Get-JCSystemUser: 101+ Users
    * [Using /systems/{system_id}/users endpoint](https://docs.jumpcloud.com/api/2.0/index.html#operation/graph_systemTraverseUser)
* Get-JCSystemGroupMember: 101+ Systems
    * [Using /Systemgroups/{group_id}/members endpoint](https://docs.jumpcloud.com/api/2.0/index.html#operation/graph_systemGroupMembersList)
* Get-JCUserGroupMember: 101+ Users
    * [Using /Usergroups/{group_id}/members endpoint](https://docs.jumpcloud.com/api/2.0/index.html#operation/graph_userGroupMembersList)

> :memo: **Note:** Even if the -parallel flag is set to $true and the above limits are not met, the function will default to running in serial

## How should this be tested?
Ensure that your testing organization meets the following requirements
* Get-JCUser: More than 1001 users
* Get-JCSystem: More than 1001 systems
* Get-JCSystemUser: More than 101 Users assigned to a device
* Get-JCSystemGroupMember: More than 101 Systems assigned to a System/Device Group
* Get-JCUserGroupMember: More than 101 Users assigned to a User Group

Pull this branch and use the following code example to compare results:
```powershell
Import-Module "/Users/USERNAME/Documents/GitHub/support/PowerShell/JumpCloud Module/JumpCloud.psd1" -force
Connect-JCOnline "APIKEY" -force
#Parallel test
Write-Host "Starting parallel..."
$ParallelUsers = Get-JCUser -parallel $true
$ParallelSystems = Get-JCSystem -parallel $true
$ParallelSystemUsers = Get-JCSystemUser -SystemID "SYSTEMID" -parallel $true
$ParallelSystemGroupMembers = Get-JCSystemGroupMember -ByID 'SYSTEMGROUPID' -parallel $true
$ParallelUserGroupMembers = Get-JCUserGroupMember -ByID "USERGROUPID" -parallel $true
Write-Host "Parallel done..."

#Serial test
Write-Host "Starting serial..."
$SerialUsers = Get-JCUser
$SerialSystems = Get-JCSystem
$SerialSystemUsers = Get-JCSystemUser -SystemID "SYSTEMID"
$SerialSystemGroupMembers = Get-JCSystemGroupMember -ByID 'SYSTEMGROUPID'
$SerialUserGroupMembers = Get-JCUserGroupMember -ByID "USERGROUPID"
Write-Host "Serial done..."

#Compare results
$ParallelUsers.Count | Should -Be $SerialUsers.Count
$SortedParallelUsers = $ParallelUsers.username | Sort-Object
$SortedSerialUsers = $SerialUsers.username | Sort-Object
$SortedParallelUsers | Should -Be $SortedSerialUsers

$ParallelSystems.Count | Should -Be $SerialSystems.Count
$SortedParallelSystems = $ParallelSystems.displayName | Sort-Object
$SortedSerialSystems = $SerialSystems.displayName | Sort-Object
$SortedParallelSystems | Should -Be $SortedSerialSystems

$ParallelSystemUsers.Count | Should -Be $SerialSystemUsers.Count
$SortedParallelSysUsers = $ParallelSystemUsers.username | Sort-Object
$SortedSerialSysUsers = $SerialSystemUsers.username | Sort-Object
$SortedParallelSysUsers | Should -Be $SortedSerialSysUsers

$ParallelSystemGroupMembers.Count | Should -Be $SerialSystemGroupMembers.Count
$SortedParallelMembers = $ParallelSystemGroupMembers | Sort-Object -Property SystemID | ConvertTo-Json
$SortedSerialMembers = $SerialSystemGroupMembers | Sort-Object -Property SystemID | ConvertTo-Json
$SortedParallelMembers | Should -Be $SortedSerialMembers

$ParallelUserGroupMembers.Count | Should -Be $SerialUserGroupMembers.Count
$SortedParallelMembers = $ParallelUserGroupMembers | Sort-Object -Property UserID | ConvertTo-Json
$SortedSerialMembers = $SerialUserGroupMembers | Sort-Object -Property UserID | ConvertTo-Json
$SortedParallelMembers | Should -Be $SortedSerialMembers
```
